web/react/components/more_direct_channels.jsx

```diff
@@ -3,67 +3,107 @@
         });
     },
     getInitialState: function() {
-        return { channels: [] };
+        return {channels: [], loadingPMChannel: -1};
     },
     render: function() {
         var self = this;
 
-        var directMessageItems = this.state.channels.map(function(channel) {
-            var badge = "";
-            var titleClass = ""
+        var directMessageItems = this.state.channels.map(function mapActivityToChannel(channel, index) {
+            var badge = '';
+            var titleClass = '';
+            var active = '';
+            var handleClick = null;
 
             if (!channel.fake) {
-                var active = channel.id === ChannelStore.getCurrentId() ? "active" : "";
+                if (channel.id === ChannelStore.getCurrentId()) {
+                    active = 'active';
+                }
 
                 if (channel.unread) {
-                    badge = <span className="badge pull-right small">{channel.unread}</span>;
-                    badgesActive = true;
-                    titleClass = "unread-title"
+                    badge = <span className='badge pull-right small'>{channel.unread}</span>;
+                    titleClass = 'unread-title';
                 }
-                return (
-                    <li key={channel.name} className={active}><a className={"sidebar-channel " + titleClass} href="#" onClick={function(e){e.preventDefault(); utils.switchChannel(channel, channel.teammate_username); $(self.refs.modal.getDOMNode()).modal('hide')}}>{badge}{channel.display_name}</a></li>
-                );
+
+                handleClick = function clickHandler(e) {
+                    e.preventDefault();
+                    utils.switchChannel(channel, channel.teammate_username);
+                    $(self.refs.modal.getDOMNode()).modal('hide');
+                };
             } else {
-                return (
-                    <li key={channel.name} className={active}><a className={"sidebar-channel " + titleClass} href={TeamStore.getCurrentTeamUrl() + "/channels/"+channel.name}>{badge}{channel.display_name}</a></li>
-                );
+                // It's a direct message channel that doesn't exist yet so let's create it now
+                var ids = channel.name.split('__');
+                var otherUserId = '';
+                if (ids[0] === UserStore.getCurrentId()) {
+                    otherUserId = ids[1];
+                } else {
+                    otherUserId = ids[0];
+                }
+
+                if (self.state.loadingPMChannel === index) {
+                    badge = <img className='channel-loading-gif pull-right' src='/static/images/load.gif'/>;
+                }
+
+                handleClick = function clickHandler(e) {
+                    e.preventDefault();
+                    self.setState({loadingPMChannel: index});
+
+                    Client.createPMChannelIfNotExists(channel, otherUserId,
```

Comment:
minor nit, but why not createDirectChannel().  I think the backend refers to these as direct channels vs private channels (which is labeled a private group in the UI).

there is:
public channel
private channel
direct channel (1:1)